### PR TITLE
Add JSON Schema validation to Lua

### PR DESCRIPTION
**Pull request by javierguerragiraldez**
New 'go.jsonschema' module, exports two functions:

`compile( <schematext> [, <draftname>] )`

Where `<schematext>` is a string containing a JSONSchema and `<draftname>`
is one of "draft4", "draft6", "draft7", "draft2019" or "draft2020".
If omitted, "draft4" is assumed.

If successful, a compiled schema object will be returned, on error it
returns nil and an error message.

`validate( <schema>, <input> )`

Where `<schema>` is a compiled schema object and `<input>` is some JSON
text.

If succesful, returns a single true value.  If the validation failes,
returns false and a descriptive error.  On any other error, returns nil
and an error message.

### CODE REVIEW

**Comment by hbagdi** [comment]
What is the reason for not being able to use the Lua library for this? Is it using a dynamic shared object?

**Comment by javierguerragiraldez** [comment]
> What is the reason for not being able to use the Lua library for this? Is it using a dynamic shared object?

from a quick review of [ljsonschema](https://github.com/Tieske/lua-resty-ljsonschema), i saw:
- it only supports Draft4 (i don't know if newer schemas can be added or if they would need updates in Lua code)
- it generates Lua code, which works great in LuaJIT, but could get heavy on pure interpreters.
- it's written for OpenResty, relying on a few libraries that are easy to replace, but have different corner cases
- it does expect cjson behaviour specifically, mostly around ambiguities in Lua about arrays vs objects.

File: lualibs/go/jsonschema/api.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+		l.Push(lua.LFalse)
+		l.Push(lua.LString(err.Error()))
+	}
+
+	return pushErr(l, err.Error())
+}
+
+func unmarshal(s string) (interface{}, error) {
+	decoder := json.NewDecoder(strings.NewReader(s))
+	decoder.UseNumber()
```
Comment (hbagdi) [inline lualibs/go/jsonschema/api.go:103]: Why is this being done? Does Lua's cjson decode numbers as string?

File: lualibs/go/jsonschema/api.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	var doc interface{}
+	if err := decoder.Decode(&doc); err != nil {
+		return nil, err
+	}
+	if t, _ := decoder.Token(); t != nil {
```
Comment (hbagdi) [inline lualibs/go/jsonschema/api.go:108]: Should this not always return an error == io.EOF?

File: lualibs/go/jsonschema/api.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	switch err := err.(type) {
+	case *jsonschema.ValidationError,
+		*jsonschema.InvalidJSONTypeError,
+		*jsonschema.InfiniteLoopError:
+
+		l.Push(lua.LFalse)
+		l.Push(lua.LString(err.Error()))
+	}
```
Comment (hbagdi) [inline lualibs/go/jsonschema/api.go:96]: Is the code missing a return here?

**Review by hbagdi — APPROVED** [review]

**Pull request merged**